File: python/nedb/log.py
# SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
# SPDX-License-Identifier: BUSL-1.1
# aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

"""
nedb.log — the append-only, hash-chained, nonce-enforced, idempotent operation log.

This is the single source of truth for NEDB. Every mutation in the database is an
Op appended here. Three guarantees live in this one structure:

  * Replay protection  — each client has a strictly-monotonic nonce; an op whose
                         nonce is <= the client's last seen nonce is rejected.
  * Idempotency        — an op carrying an idempotency key that was already applied
                         returns the original result and is NOT appended again.
  * Tamper evidence    — ops are chained by hash (h_n = H(h_{n-1} || op_n)), so the
                         whole history is a verifiable chain and the head hash is a
                         commitment to the entire log (anchorable on a blockchain).

The same log is the substrate for MVCC snapshot isolation, crash recovery, and
time-travel reads: every Op has a monotonic `seq`, and state "AS OF seq N" is just
the log truncated at N.
"""
from __future__ import annotations

import hashlib
import json
import time
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

GENESIS = "0" * 64


def canon(obj: Any) -> bytes:
    """Deterministic canonical encoding for hashing."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), default=str).encode()


def blake(data: bytes) -> str:
    # Reference uses BLAKE2b (stdlib). The production Rust core uses BLAKE3
    # (faster, natively tree-structured for the Merkle history).
    return hashlib.blake2b(data, digest_size=32).hexdigest()


class ReplayError(Exception):
    """Raised when an op is replayed with a stale/duplicate nonce."""


@dataclass
class Op:
    seq: int
    client: str
    nonce: int
    op: str  # put | delete | link | unlink | put_file
    payload: dict
    ts: float
    idem: Optional[str]
    prev_hash: str
    hash: str
    # ── Causal provenance (v0.9.0+) ─────────────────────────────────────────
    # Optional fields that, when present, are sealed inside the hash chain so
    # they are tamper-evident and time-stamped at write time.
    #   caused_by  — seqs of the ops that led to this write (backward trace).
    #   evidence   — source type: "user_message" | "inference" | "tool_result"
    #                             | "correction" | "external"
    #   confidence — agent's certainty in this write (0.0 – 1.0).
    caused_by:  Optional[List[int]] = None
    evidence:   Optional[str]       = None
    confidence: Optional[float]     = None

    # ── Bi-temporal valid time (v1.0.0+) ─────────────────────────────────────
    # When was this fact TRUE IN THE WORLD (independent of when it was written)?
    #   valid_from — ISO 8601 date/datetime string; None = "from the beginning"
    #   valid_to   — ISO 8601 date/datetime string; None = "still valid / open-ended"
    #
    # ISO 8601 strings sort lexicographically correctly, so comparisons are
    # safe as plain string ops:  "2024-01-01" < "2024-06-15" ✓
    #
    # Backward-compatible: ops without valid-time fields are treated as always
    # valid (they pass every VALID AS OF filter). Existing chains verify unchanged.
    valid_from: Optional[str] = None
    valid_to:   Optional[str] = None

    def to_dict(self) -> dict:
        """Serialize for the append-only log file (AOF)."""
        d: dict = {
            "seq": self.seq, "client": self.client, "nonce": self.nonce,
            "op": self.op, "payload": self.payload, "ts": self.ts,
            "idem": self.idem, "prev_hash": self.prev_hash, "hash": self.hash,
        }
        if self.caused_by  is not None: d["caused_by"]  = self.caused_by
        if self.evidence   is not None: d["evidence"]   = self.evidence
        if self.confidence is not None: d["confidence"] = self.confidence
        if self.valid_from is not None: d["valid_from"] = self.valid_from
        if self.valid_to   is not None: d["valid_to"]   = self.valid_to
        return d

    @classmethod
    def from_dict(cls, d: dict) -> "Op":
        return cls(
            d["seq"], d["client"], d["nonce"], d["op"], d["payload"],
            d["ts"], d.get("idem"), d["prev_hash"], d["hash"],
            caused_by  = d.get("caused_by"),
            evidence   = d.get("evidence"),
            confidence = d.get("confidence"),
            valid_from = d.get("valid_from"),
            valid_to   = d.get("valid_to"),
        )


class OpLog:
    def __init__(self) -> None:
        self.ops: List[Op] = []
        self._last_nonce: Dict[str, int] = {}
        self._idem: Dict[str, int] = {}  # idem key -> seq of original op
        self._head = GENESIS

    def append(
        self,
        client: str,
        nonce: int,
        op: str,
        payload: dict,
        idem: Optional[str] = None,
        ts: Optional[float] = None,
        caused_by:  Optional[List[int]] = None,
        evidence:   Optional[str]       = None,
        confidence: Optional[float]     = None,
        valid_from: Optional[str]       = None,
        valid_to:   Optional[str]       = None,
    ) -> Tuple[Op, bool]:
        """Append an op. Returns (op, created). `created` is False when the op was
        deduplicated by its idempotency key (a no-op replay-safe return)."""
        # Idempotency: a known key returns the original op without re-appending.
        if idem is not None and idem in self._idem:
            return self.ops[self._idem[idem]], False

        # Replay protection: nonce must strictly exceed the client's last nonce.
        last = self._last_nonce.get(client, 0)
        if nonce <= last:
            raise ReplayError(
                f"replay/stale nonce for client '{client}': {nonce} <= {last}"
            )

        seq = len(self.ops)
        ts = time.time() if ts is None else ts
        body: dict = {
            "seq": seq, "client": client, "nonce": nonce,
            "op": op, "payload": payload, "ts": ts, "idem": idem,
        }
        # Provenance fields are sealed INTO the hash when present so they are
        # tamper-evident — omitting them when absent keeps old ops verifiable.
        if caused_by  is not None: body["caused_by"]  = caused_by
        if evidence   is not None: body["evidence"]   = evidence
        if confidence is not None: body["confidence"] = confidence
        if valid_from is not None: body["valid_from"] = valid_from
        if valid_to   is not None: body["valid_to"]   = valid_to
        h = blake(self._head.encode() + canon(body))
        rec = Op(seq, client, nonce, op, payload, ts, idem, self._head, h,
                 caused_by=caused_by, evidence=evidence, confidence=confidence,
                 valid_from=valid_from, valid_to=valid_to)

        self.ops.append(rec)
        self._last_nonce[client] = nonce
        if idem is not None:
            self._idem[idem] = seq
        self._head = h
        return rec, True

    def load(self, ops: List[Op]) -> None:
        """Rehydrate the log from persisted ops WITHOUT recomputing hashes, so the
        original chain (and thus verify() and the head commitment) is preserved
        exactly across a restart. Nonce, idempotency, and head state are restored
        from the ops themselves — replay protection survives a reload."""
        self.ops = list(ops)
        self._last_nonce = {}
        self._idem = {}
        for o in self.ops:
            if o.nonce > self._last_nonce.get(o.client, 0):
                self._last_nonce[o.client] = o.nonce
            if o.idem is not None and o.idem not in self._idem:
                self._idem[o.idem] = o.seq
        self._head = self.ops[-1].hash if self.ops else GENESIS

    @staticmethod
    def _op_body(o: "Op") -> dict:
        """The canonical hash body for an op — must match exactly what append() hashes."""
        body: dict = {
            "seq": o.seq, "client": o.client, "nonce": o.nonce,
            "op": o.op, "payload": o.payload, "ts": o.ts, "idem": o.idem,
        }
        # Optional fields included only when present (backward-compat with old ops).
        if o.caused_by  is not None: body["caused_by"]  = o.caused_by
        if o.evidence   is not None: body["evidence"]   = o.evidence
        if o.confidence is not None: body["confidence"] = o.confidence
        if o.valid_from is not None: body["valid_from"] = o.valid_from
        if o.valid_to   is not None: body["valid_to"]   = o.valid_to
        return body

    def verify(self) -> bool:
        """Re-walk the chain and confirm no op has been tampered with."""
        prev = GENESIS
        for o in self.ops:
            body = self._op_body(o)
            if o.prev_hash != prev:
                return False
            if o.hash != blake(prev.encode() + canon(body)):
                return False
            prev = o.hash
        return True

    @property
    def head(self) -> str:
        return self._head

    def slice_until(self, as_of: int) -> List[Op]:
        return [o for o in self.ops if o.seq <= as_of]

    def __len__(self) -> int:
        return len(self.ops)

File: rust/crates/nedb-core/src/log.rs
// SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
// SPDX-License-Identifier: BUSL-1.1
// aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

//! The append-only, hash-chained, nonce-enforced, idempotent operation log.
//!
//! This is the single source of truth. Replay protection (strictly-monotonic
//! per-client nonces), idempotency (dedup by key), and tamper evidence (BLAKE3
//! hash chain) all live here. The head hash commits to the entire history and is
//! anchorable on-chain (e.g. ITC).

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const GENESIS: &str =
    "0000000000000000000000000000000000000000000000000000000000000000";

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Op {
    pub seq: u64,
    pub client: String,
    pub nonce: u64,
    pub op: String,
    pub payload: Value,
    pub idem: Option<String>,
    pub prev_hash: String,
    pub hash: String,
}

#[derive(Debug, Clone)]
pub enum LogError {
    /// An op was replayed with a stale/duplicate nonce.
    Replay { client: String, nonce: u64, last: u64 },
}

impl std::fmt::Display for LogError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LogError::Replay { client, nonce, last } => write!(
                f,
                "replay/stale nonce for client '{client}': {nonce} <= {last}"
            ),
        }
    }
}
impl std::error::Error for LogError {}

/// Canonical JSON encoding for hashing. `serde_json::Value::Object` is backed by a
/// BTreeMap by default, so keys serialize in sorted order — deterministic across
/// platforms and language bindings.
fn canonical(v: &Value) -> String {
    serde_json::to_string(v).expect("serialize op body")
}

fn chain_hash(prev: &str, body: &Value) -> String {
    let mut h = blake3::Hasher::new();
    h.update(prev.as_bytes());
    h.update(canonical(body).as_bytes());
    h.finalize().to_hex().to_string()
}

#[derive(Default)]
pub struct OpLog {
    pub ops: Vec<Op>,
    last_nonce: HashMap<String, u64>,
    idem: HashMap<String, usize>,
    head: String,
}

impl OpLog {
    pub fn new() -> Self {
        Self {
            ops: Vec::new(),
            last_nonce: HashMap::new(),
            idem: HashMap::new(),
            head: GENESIS.to_string(),
        }
    }

    /// Append an op. Returns `(op, created)`; `created == false` means the op was
    /// deduplicated by its idempotency key (replay-safe no-op return).
    pub fn append(
        &mut self,
        client: &str,
        nonce: u64,
        op: &str,
        payload: Value,
        idem: Option<String>,
    ) -> Result<(Op, bool), LogError> {
        if let Some(k) = &idem {
            if let Some(&seq) = self.idem.get(k) {
                return Ok((self.ops[seq].clone(), false));
            }
        }
        let last = *self.last_nonce.get(client).unwrap_or(&0);
        if nonce <= last {
            return Err(LogError::Replay {
                client: client.to_string(),
                nonce,
                last,
            });
        }

        let seq = self.ops.len() as u64;
        let body = serde_json::json!({
            "seq": seq, "client": client, "nonce": nonce,
            "op": op, "payload": payload, "idem": idem,
        });
        let hash = chain_hash(&self.head, &body);
        let rec = Op {
            seq,
            client: client.to_string(),
            nonce,
            op: op.to_string(),
            payload,
            idem: idem.clone(),
            prev_hash: std::mem::replace(&mut self.head, hash.clone()),
            hash,
        };
        self.ops.push(rec.clone());
        self.last_nonce.insert(client.to_string(), nonce);
        if let Some(k) = idem {
            self.idem.insert(k, seq as usize);
        }
        Ok((rec, true))
    }

    /// Re-walk the chain and confirm no op has been tampered with.
    pub fn verify(&self) -> bool {
        let mut prev = GENESIS.to_string();
        for o in &self.ops {
            if o.prev_hash != prev {
                return false;
            }
            let body = serde_json::json!({
                "seq": o.seq, "client": o.client, "nonce": o.nonce,
                "op": o.op, "payload": o.payload, "idem": o.idem,
            });
            if o.hash != chain_hash(&prev, &body) {
                return false;
            }
            prev = o.hash.clone();
        }
        true
    }

    pub fn head(&self) -> &str {
        &self.head
    }

    pub fn len(&self) -> usize {
        self.ops.len()
    }

    pub fn is_empty(&self) -> bool {
        self.ops.is_empty()
    }

    /// Rehydrate from a slice of persisted ops WITHOUT recomputing hashes.
    /// The original chain (and thus verify() and the head commitment) is preserved.
    pub fn load(&mut self, ops: Vec<Op>) {
        self.ops = ops;
        self.last_nonce.clear();
        self.idem.clear();
        for (i, o) in self.ops.iter().enumerate() {
            let prev = self.last_nonce.get(&o.client).copied().unwrap_or(0);
            if o.nonce > prev {
                self.last_nonce.insert(o.client.clone(), o.nonce);
            }
            if let Some(k) = &o.idem {
                self.idem.entry(k.clone()).or_insert(i);
            }
        }
        self.head = self.ops.last().map(|o| o.hash.clone()).unwrap_or_else(|| GENESIS.to_string());
    }

    /// Export the current nonce map (for persistence restore).
    pub fn nonce_map(&self) -> HashMap<String, u64> {
        self.last_nonce.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn replay_is_rejected() {
        let mut log = OpLog::new();
        log.append("a", 1, "put", serde_json::json!({}), None).unwrap();
        let err = log.append("a", 1, "put", serde_json::json!({}), None);
        assert!(err.is_err());
    }

    #[test]
    fn idempotent_key_dedups() {
        let mut log = OpLog::new();
        let (_, c1) = log
            .append("a", 1, "put", serde_json::json!({}), Some("k".into()))
            .unwrap();
        let (_, c2) = log
            .append("a", 2, "put", serde_json::json!({}), Some("k".into()))
            .unwrap();
        assert!(c1 && !c2);
        assert_eq!(log.len(), 1);
    }

    #[test]
    fn chain_verifies_and_detects_tamper() {
        let mut log = OpLog::new();
        log.append("a", 1, "put", serde_json::json!({"x": 1}), None)
            .unwrap();
        log.append("a", 2, "put", serde_json::json!({"x": 2}), None)
            .unwrap();
        assert!(log.verify());
        log.ops[0].payload = serde_json::json!({"x": 999});
        assert!(!log.verify());
    }
}
